govc/vm/clone.go

```diff
@@ -316,6 +320,49 @@ func (cmd *clone) cloneVM(ctx context.Context) (*object.Task, error) {
 		Template: cmd.template,
 	}
 
+	if cmd.link == "snapshot" {
+		relocateSpec.DiskMoveType = string(types.VirtualMachineRelocateDiskMoveOptionsCreateNewChildDiskBacking)
+		cloneSpec.Location = relocateSpec
+		//get snapshots:
+		var o mo.VirtualMachine
+		vm := cmd.VirtualMachine
+		err = vm.Properties(ctx, vm.Reference(), []string{"snapshot"}, &o)
+		if err != nil {
+			return nil, err
+		}
+
+		if o.Snapshot == nil {
+			return nil, fmt.Errorf("Given VM does not have a snapshot to clone from. Create one, or use -link=machine to create a machine linked clone")
+		}
+
+		//make sure we have a snapshot, and set current snapshot as link source
+		if o.Snapshot.CurrentSnapshot == nil {
+			if cmd.snapshot == "" {
+				return nil, fmt.Errorf("Given VM does not have a current snapshot to clone from, specify a snapshot name using -snapshot=")
+			}
+		} else {
+			//default to current snapshot
+			cloneSpec.Snapshot = o.Snapshot.CurrentSnapshot
+		}
+
+		//if we should use a specific snapshot
+		if cmd.snapshot != "" && cmd.snapshot != "current" {
+			snap, err := vm.FindSnapshot(ctx, cmd.snapshot)
+			if err != nil {
+				return nil, err
+			}
+
+			if snap == nil {
```

Comment:
Note that `FindSnapshot` does not return an nil snapshot without having returned a non-nil error, so there's no need for this check.

Reply:
I suppose to keep it that way, we'd probably want to change the diff above to:
```diff
@@ -534,6 +534,14 @@ func (v VirtualMachine) FindSnapshot(ctx context.Context, name string) (Referenc
 		return nil, errors.New("No snapshots for this VM")
 	}
 
+	if name == "" {
+		if o.Snapshot.CurrentSnapshot != nil {
+			return o.Snapshot.CurrentSnapshot, nil
+		}
+
+		return nil, errors.New("No current snapshot for this VM")
+	}
+
 	m := make(snapshotMap)
 	m.add("", o.Snapshot.RootSnapshotList)
 
```